[GlobalISel] convergence control tokens and intrinsics

ssahasra opened the pull request:
Support for lowering convergence control to GMIR:
- Introduce new G_CONVERGENCECTRL_* opcodes.
- In the IR translator, convert the LLVM token type to LLT::token(), which is an alias for the s0 type. These show up as implicit uses on convergent operations.
- In the machine verifier, enforce the static rules of convergence control.

Note that the lowering of the new GMIR opcodes is entirely target-specific. It is generally expected that the backend will use convergence control to change the CFG of each function, and then discard the tokens. This is currently a work in progress for AMDGPU.

Differential Revision: https://reviews.llvm.org/D158147

nhaehnle reviewed (commented):
I haven't had the time to review this in detail, but the overall approach seems sound to me.

arsenm commented on llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:
```diff
@@ -121,6 +132,8 @@ bool CallLowering::lowerCall(MachineIRBuilder &MIRBuilder, const CallBase &CB,
     CanBeTailCalled = false;
   }
 
+  if (!hasConvergenceEntryToken(CB))
```
Should this go under a CanBeTailCalled check?

arsenm replied:
Don't understand this part; if there's any relation between tail call and convergence I would expect it to be the opposite

arsenm replied:
Ping? Why is this here and/or backwards?

ssahasra replied:
It's just plain wrong. I no longer remember what I was doing when I put it in, but I have now implemented support for tokens on tail calls. Will push it along with responses to the other comments.

ssahasra committed:
[GlobalISel] convergence control tokens and intrinsics

In the IR translator, convert the LLVM token type to LLT::token(), which is an
alias for the s0 type. These show up as implicit uses on convergent operations.

Differential Revision: https://reviews.llvm.org/D158147

ssahasra commented:
- Rebased on top of main, which includes the SelectionDAG changes for convergence tokens.
- Eliminated the G_* opcodes in favour of the opcodes introduced for SelectionDAG

ssahasra commented on llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-convergence-tokens.ll:
This file got leftover from the previous commit. It's not necessary because the adjacent file "convergence-tokens.ll" now checks for GlobalISel too.

arsenm commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -2036,6 +2037,37 @@ bool IRTranslator::translateIfEntryValueArgument(bool isDeclare, Value *Val,
+  switch (ID) {
+  default:
+    llvm_unreachable("Unexpected intrinsic");
+    return 0;
```
No dead return

arsenm commented on llvm/include/llvm/CodeGenTypes/LowLevelType.h:
```diff
@@ -45,6 +45,14 @@ class LLT {
   }
 
+  /// Get a low-level token; just a scalar with zero bits (or no size).
+  static constexpr LLT token() {
```
This should get some new tests in unittests/CodeGen/LowLevelTypeTest.cpp

ssahasra committed:
[GlobalISel] Introduce LLT:token() as a special scalar type

The new token type is used in #67006 for implementing convergence control tokens
in GMIR.

ssahasra commented:
- rebased to upstream/main
- split out LLT::token() as a separate commit
- added support for tail calls along with a test
- removed redundant return (after an unreachable)

ssahasra committed:
Merge branch 'main' into ssahasra/gmir-convergence-tokens
